src/hotspot/share/cds/filemap.cpp

```diff
@@ -842,6 +885,68 @@ bool FileMapInfo::classpath_failure(const char* msg, const char* name) {
   return false;
 }
 
+char* FileMapInfo::longest_common_app_classpath(int num_paths,
+                                                GrowableArray<const char*>* rp_array) {
+  // find the shortest path length among all the entries
+  char* chr = strrchr((char*)rp_array->at(0), *os::file_separator());
+  size_t min_len = chr != NULL ? chr - rp_array->at(0) + 1 : strlen(rp_array->at(0)) + 1;
+  for (int i = 1; i < num_paths; i++) {
+    chr = strrchr((char*)rp_array->at(i), *os::file_separator());
+    size_t curr_len = chr != NULL ? chr - rp_array->at(i) + 1 : strlen(rp_array->at(i)) + 1;
+    min_len = MIN2(min_len, curr_len);
+  }
+
+  // allocate a buffer to store the longest common prefix
+  char* lcp = NEW_C_HEAP_ARRAY(char, min_len + 1, mtInternal);
+  memset(lcp, 0, min_len + 1);
+  char c;
+  // find the longest common prefix
+  for (int j = 0; j < (int)min_len; j++) {
+    c = rp_array->at(0)[j];
+    int k = 1;
+    while (k < num_paths) {
+      if (rp_array->at(k)[j] != c) {
+        break;
+      }
+      k++;
+    }
+    *(lcp + j) = c;
+  }
+
```

Comment:
Need to check for this case as well:

```
rp[0] = "/x/a/Foo.jar"
rp[1] = "/x/aa/Bar.jar"
```

I think the current code will incorrectly return "/x/a"

Reply:
Fixed. I rewrote the function. It now returns the length.
I also added some test cases to test the above and refactored some code in the test cases.